flink-tests/src/test/java/org/apache/flink/test/checkpointing/RegionFailoverITCase.java

```diff
@@ -0,0 +1,307 @@
+import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.state.CheckpointListener;
+import org.apache.flink.runtime.state.FunctionInitializationContext;
+import org.apache.flink.runtime.state.FunctionSnapshotContext;
+import org.apache.flink.runtime.testutils.MiniClusterResourceConfiguration;
+import org.apache.flink.streaming.api.CheckpointingMode;
+import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction;
+import org.apache.flink.streaming.api.environment.CheckpointConfig;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.functions.source.RichParallelSourceFunction;
+import org.apache.flink.test.util.MiniClusterWithClientResource;
+import org.apache.flink.util.TestLogger;
+
+import org.apache.commons.collections.CollectionUtils;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.StreamSupport;
+
+/**
+ * Tests for region failover with multi regions.
+ */
+public class RegionFailoverITCase extends TestLogger {
+
+	private static final int FAIL_BASE = 1000;
+	private static final int NUM_OF_REGIONS = 3;
+	private static final Set<Integer> EXPECTED_INDICES = IntStream.range(0, NUM_OF_REGIONS).boxed().collect(Collectors.toSet());
+	private static final int NUM_OF_RESTARTS = 3;
+	private static final int NUM_ELEMENTS = FAIL_BASE * 10;
+
+	private static volatile long lastCompletedCheckpointId = 0;
+	private static volatile int numCompletedCheckpoints = 0;
+	private static volatile AtomicInteger jobFailedCnt = new AtomicInteger(0);
+
+	private static Map<Long, Integer> snapshotIndicesOfSubTask = new HashMap<>();
+
+	private static MiniClusterWithClientResource cluster;
+
+	private static boolean restoredState = false;
+
+	@ClassRule
+	public static final TemporaryFolder TEMPORARY_FOLDER = new TemporaryFolder();
+
+	@Before
+	public void setup() throws Exception {
+		Configuration configuration = new Configuration();
+		configuration.setString(JobManagerOptions.EXECUTION_FAILOVER_STRATEGY, "region");
+
+		cluster = new MiniClusterWithClientResource(
+			new MiniClusterResourceConfiguration.Builder()
+				.setConfiguration(configuration)
+				.setNumberTaskManagers(2)
+				.setNumberSlotsPerTaskManager(2).build());
+		cluster.before();
+		jobFailedCnt.set(0);
+		numCompletedCheckpoints = 0;
+	}
+
+	@AfterClass
+	public static void shutDownExistingCluster() {
+		if (cluster != null) {
+			cluster.after();
+			cluster = null;
+		}
+	}
+
+	/**
+	 * Tests that a simple job (Source -> Map) with multi regions could restore with operator state.
```

Comment:
This test does not check for restore of keyed state. Maybe we can easily also include keyed state if we use `DataStreamUtils::reinterpretAsKeyedStream` and soures that make sure they follow the right subtask index assignment pattern, e.g. only produce elements with the right hash code to be routed to their downstream task.